includes/class-wporg-gp-translation-events-translation-listener.php

```diff
@@ -50,68 +56,123 @@ function ( $translation, $translation_before ) {
+
+			/** @var GP_Translation_Set $translation_set */
+			$translation_set = ( new GP_Translation_Set )->find_one( [ 'id' => $translation->translation_set_id ] );
+			global $wpdb;
+
+			foreach ( $event_ids as $event_id ) {
+				// A given user can only do one action on a specific translation.
+				// So we replace instead of insert, which will keep only the last action.
+				$wpdb->replace(
+					self::ACTIONS_TABLE_NAME,
+					[
+						// start primary key
+						'event_id'       => $event_id,
+						'user_id'        => $user_id,
+						'translation_id' => $translation->id,
+						// end primary key
+						'action'         => $action,
+						'locale'         => $translation_set->locale,
+						'happened_at'    => $happened_at->format( 'Y-m-d H:i:s' ),
+					]
+				);
+			}
+		} catch (Exception $exception) {
+			error_log($exception);
 		}
 	}
 
 	/**
-	 * @return WP_Post[]
+	 * @return int[]
+	 * @throws Exception
 	 */
-	private function get_active_events( DateTime $at ): array {
-		return get_posts(
+	private function get_active_events( DateTimeImmutable $at ): array {
+		$cache_entries = $this->active_events_cache->get();
+		if ( null === $cache_entries ) {
+			$cache_entries = $this->cache_events( $at );
+		}
+
+		// Filter out events that aren't actually active at $at.
+		$events = array_filter(
+			$cache_entries,
+			function ( $cache_entry ) use ( $at ) {
+				return $at >= $cache_entry->start && $at <= $cache_entry->end;
+			}
+		);
+
+		return array_map(
+			function ( $cache_entry ) {
+				return $cache_entry->event_id;
+			},
+			$events
+		);
+	}
+
+	/**
+	 * @return WPORG_GP_Translation_Events_Active_Events_Cache_Entry[]
+	 * @throws Exception
+	 */
+	private function cache_events( DateTimeImmutable $at ): array {
+		$cache_duration = WPORG_GP_Translation_Events_Active_Events_Cache::CACHE_DURATION;
+		$boundary_start = $at;
+		$boundary_end   = $at->modify( "+$cache_duration seconds" );
+
+		// Get events for which start is before $boundary_end AND end is after $boundary_start.
+		$event_ids = get_posts(
 			[
-				'post_type'   => 'event',
-				'post_status' => 'publish',
-				'meta_query'  => [
+				'post_type'      => 'event',
+				'post_status'    => 'publish',
+				'posts_per_page' => -1,
+				'fields'         => 'ids',
+				'meta_query'     => [
 					[
 						'key'     => '_event_start',
-						'value'   => $at->format( 'Y-m-d H:i:s' ),
-						'compare' => '<=',
+						'value'   => $boundary_end->format( 'Y-m-d H:i:s' ),
+						'compare' => '<',
 						'type'    => 'DATETIME',
 					],
 					[
 						'key'     => '_event_end',
-						'value'   => $at->format( 'Y-m-d H:i:s' ),
-						'compare' => '>=',
+						'value'   => $boundary_start->format( 'Y-m-d H:i:s' ),
+						'compare' => '>',
 						'type'    => 'DATETIME',
 					],
 				],
 			],
 		);
+
+		$cache_entries = [];
+		foreach ( $event_ids as $event_id ) {
+			$meta = get_post_meta( $event_id );
```

Comment:
Is there a way to get the post meta of multiple posts in one query?

Reply:
AFAIK just with a custom query like
```
    $sql = $wpdb->prepare(
        "SELECT post_id, meta_value FROM $wpdb->postmeta WHERE meta_key = %s AND post_id IN ( $post_ids )",
        $meta_key
    );

```

Reply:
Do you think it's worth implementing this, or calling `get_post_meta()` for each event would be fine? This code will only run when the cache is empty, so once per day, or when events change.

Reply:
It's find to use `get_post_meta` for now. We can always optimize later.